blocks/blocks.js

```diff
@@ -359,18 +368,242 @@ class Blocks extends Newsletter {
         this._dragSvg.removeChild(script.elem)
         if (possibleDropTarget && possibleDropTarget.acceptDrop) {
           const { x, y } = script.position
-          possibleDropTarget.acceptDrop(
+          undoEntry.b = possibleDropTarget.acceptDrop(
             script,
             x,
             y,
             snapTo,
-            wrappingC
+            wrappingC,
+            undoEntry
           )
+        }
+        if (!undoEntry.b) {
+          undoEntry.b = script.toJSON().blocks
+        }
+        const extraSteps = this.shoveTarget(undoEntry.b, script)
+        this.addUndoEntry(extraSteps ? [...extraSteps, undoEntry] : undoEntry)
+      }
+    }
+  }
+
+  /**
+   * This gets the target from the given data, displacing it from its original
+   * home if necessary.
+   * @param {number} blockCount - Maximum number of blocks to take (for
+   *   taking from within a stack - required)
+   * @returns {Script}
+   */
+  grabTarget (data, blockCount = 0) {
+    if (Array.isArray(data)) {
+      // Is an array of block JSONs (implying it is a concept to be realized)
+      const script = this.scriptFromJSON({ x: 0, y: 0, blocks: data })
+      script.resize()
+      return script
+    } else if (data.indices) {
+      // Probably means the block was dragged out from a script such that the
+      // old script still remains.
+      const [workspace, scriptIndex, ...indices] = data.indices
+      let component = workspace.scripts[scriptIndex]
+      for (const index of indices) {
+        if (component instanceof Input) {
+          component = component.getValue()
+          if (!(component instanceof Block)) {
+            throw new Error('hwat. The indices point to an input that does not hold a block.')
+          }
+        }
+        if (typeof index === 'number') {
+          component = component.components[index]
+        } else {
+          component = component.getParamComponent(index)
+        }
+      }
```

Comment:
This is basically the same as the thing in `shoveTarget`; could shove this into a helper method?